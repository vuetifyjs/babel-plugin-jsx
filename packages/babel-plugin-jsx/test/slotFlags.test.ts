import { transpile } from './util';

// stable slots
test.for([
  {
    name: 'basic',
    source: `<A>
      {{ default: () => {} }}
    </A>`,
  },
])('stable slots $name', async ({ source }) => {
  expect(await transpile(source, { optimize: true })).toMatchSnapshot();
});

// dynamic slots
test.for([
  {
    name: 'nested slots',
    source: `<A>
      {{ foo: ({ bar }) => (
        <B>
          {{ foo: () => {
            // dynamic
            return <>{ bar }</>
          }}}
        </B>
      )}}
    </A>`,
  },
  {
    name: 'computed key',
    source: `<A>
      {{
        [name]: () => {},
      }}
    </A>`,
  },
  {
    name: 'conditional value',
    source: `<A>
      {{
        default: foo ? () => {} : undefined,
      }}
    </A>`,
  },
])('dynamic slots $name', async ({ source }) => {
  expect(await transpile(source, { optimize: true })).toMatchSnapshot();
});

// forwarded slots
test.for([
  {
    name: 'simple forward',
    source: `<A>
      {{ foo: slots.foo }}
    </A>`,
  },
  {
    skip: true, // TODO: need to analyze the function for usages
    name: 'used in slot function',
    source: `<A>
      {{ foo: () => slots.foo?.() }}
    </A>`,
  },
  {
    skip: true, // TODO: need to analyze the function for usages
    name: 'deep used in slot function',
    source: `<A>
      {{ foo: () => (
        <>
          { slots.foo() }
        </>
      )}}
    </A>`,
  },
  {
    name: 'spread all',
    source: `<A>
      {{ ...slots }}
    </A>`,
  },
  {
    name: 'v-slots passthrough',
    source: `<A v-slots={ slots } />`,
  },
  {
    name: 'spread and add',
    source: `<A>
      {{ 
        ...slots,
        foo: () => {},
      }}
    </A>`,
  },
  {
    name: 'conditional on slot existence',
    source: `<A>
      {{ default: slots.default ? slots.default() : undefined }}
    </A>`,
  },
])('forwarded slots $name', async ({ source, skip }, ctx) => {
  if (skip) ctx.skip();
  expect(await transpile(source, { optimize: true })).toMatchSnapshot();
});
