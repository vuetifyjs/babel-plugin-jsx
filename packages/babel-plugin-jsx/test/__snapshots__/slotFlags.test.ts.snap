// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dynamic slots 'computed key' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  [name]: () => {},
  _: 2
}, 1024);"
`;

exports[`dynamic slots 'conditional on slot existence' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  default: slots.default ? slots.default() : () => {},
  _: 2
}, 1024);"
`;

exports[`dynamic slots 'conditional value' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  default: foo ? () => {} : undefined,
  _: 2
}, 1024);"
`;

exports[`dynamic slots 'nested slots' 1`] = `
"import { Fragment as _Fragment, createElementVNode as _createElementVNode, resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  foo: ({
    bar
  }) => _createVNode(_resolveComponent("B"), null, {
    foo: () => {
      // dynamic
      return _createElementVNode(_Fragment, null, [bar], 1024);
    },
    _: 2
  }, 1024),
  _: 1
});"
`;

exports[`forwarded slots 'conditional on slot existence' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  default: slots.default ? slots.default() : undefined,
  _: 3
});"
`;

exports[`forwarded slots 'simple forward' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  foo: slots.foo,
  _: 3
});"
`;

exports[`forwarded slots 'spread all' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  ...slots,
  _: 3
});"
`;

exports[`forwarded slots 'spread and add' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  ...slots,
  foo: () => {},
  _: 3
});"
`;

exports[`forwarded slots 'v-slots passthrough' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  ...slots,
  _: 3
});"
`;

exports[`stable slots 'basic' 1`] = `
"import { resolveComponent as _resolveComponent, createVNode as _createVNode } from "vue";
_createVNode(_resolveComponent("A"), null, {
  default: () => {},
  _: 1
});"
`;
